[WIP] HPA - pod selection strategy

omerap12 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

omerap12 committed:
[WIP] HPA - pod selection strategy

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Change to SelectionStrategy

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add converstions

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add converstions

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Fixed bug

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
mute unit tests for now

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
implement all functions

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add unit tests for pod_filters

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add feature gate

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add feature gate

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add e2e tests

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
make update

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
added adrian to kep

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey commented on pkg/features/kube_features.go:
```diff
@@ -309,6 +309,12 @@ const (
+	// kep: https://kep.k8s.io/5325
+
+	// Improve pod selection accuracy across workload types
+	HPAselectionStrategy featuregate.Feature = "HPAselectionStrategy"
```
I'm not too sure on this nit: 
```suggestion
	HPASelectionStrategy featuregate.Feature = "HPASelectionStrategy"
```

But it matches the one above, "HPAScaleToZero" now

omerap12 replied:
Yeah I thought the one above is a mistake but I don't mind changing it to align with the others

omerap12 committed:
fallback error

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
added comment

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Run make update

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Allow HPA controller access to apps resources

omerap12 committed:
Update downgrade feature gate

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Add validation for HPA's SelectionStrategy

adrianmoisey committed:
Remove unused DefaultSelectionStrategy

adrianmoisey committed:
Mark as Alpha

adrianmoisey committed:
s/HPAselectionStrategy/HPASelectionStrategy/

omerap12 committed:
make update

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
fixed unit-tests

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
fixed e2e tests

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Move podFilterForHpa() inside the computeStatusForPodsMetric() function

This matches what we do in computeStatusForResourceMetricGeneric() and to tolerancesForHpa

omerap12 committed:
Add caching to controllers

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Add podFilter

Added to:
getReadyPodsCount() - for use by ObjectMetricSourceType, when type is autoscalingv2.AverageValueMetricType
and for ExternalMetricSourceType, when type is Target.Value

omerap12 committed:
use dynamic client

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
remove app client

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
removed error logs

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
moved dynamic client to interface

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Add tests for SelectionStrategy strategy

adrianmoisey committed:
Fix test when updating status

adrianmoisey committed:
Merge branch 'master' into hpaPodSelectionStrategy

adrianmoisey committed:
fix test

adrianmoisey committed:
Should be false

omerap12 committed:
Add test for TestPrepareForUpdateSelectionStrategyEnabled func

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add tests for pod_filters and cache

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add metrics for cache hit & miss

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Add mock monitor

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Fix test

omerap12 committed:
Remove strategy status

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Revert "Remove strategy status"

This reverts commit 0796eef36d5ee58fd8bc1d2e369f6a1f7b0d7a6a.

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

omerap12 committed:
Removed CurrentSelectionStrategy from HPA status

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Update SelectionStrategy description

adrianmoisey committed:
Handle v1 round-tripping

adrianmoisey committed:
Remove status and handle v2beta1 round tripping

omerap12 committed:
Add unit tests for replica_calculator

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Remove roundtripping for v2beta1

adrianmoisey committed:
Remove v2beta2 round tripping

adrianmoisey committed:
Revert "Remove v2beta2 round tripping"

This reverts commit 061de867619b7934d71250dcd33de94b1d3a5869.

omerap12 committed:
Merge branch 'master' into hpaPodSelectionStrategy

Signed-off-by: Omer Aplatony <omerap12@gmail.com>

adrianmoisey committed:
Add other round-trip direction

And don't json Marshall it, just stringify

adrianmoisey committed:
Round-trip v2beta1
